samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -19,258 +19,212 @@
-        val checkpointPayload = ByteBuffer.wrap(msg.getMessage.asInstanceOf[Array[Byte]])
-        handleEntry(checkpointPayload, checkpointKey)
+      val checkpointKey = checkpointKeySerde.fromBytes(keyBytes)
+      val expectedGrouperFactory = checkpointKey.getGrouperFactoryClassName
+      if (!grouperFactory.equals(expectedGrouperFactory)) {
+        throw new SamzaException("SSPGrouperFactory in the checkpoint topic does not match the configured value" +
+          s"Found: $grouperFactory Expected: $expectedGrouperFactory")
       }
-    }
-    info("Done reading %s messages from checkpoint system:%s topic:%s" format(msgCount, systemName, checkpointTopic))
-  }
 
-  override def start {
-    val CHECKPOINT_STREAMID = "unused-temp-checkpoint-stream-id"
-    val spec = new KafkaStreamSpec(CHECKPOINT_STREAMID,
-                                   checkpointTopic, systemName, 1,
-                                   replicationFactor, checkpointTopicProperties)
-
-    info("About to create checkpoint stream: " + spec)
-    systemAdmin.createStream(spec)
-    info("Created checkpoint stream: " + spec)
-    try {
-      systemAdmin.validateStream(spec) // SPECIAL VALIDATION FOR CHECKPOINT. DO NOT FAIL IF failOnCheckpointValidation IS FALSE
-      info("Validated spec: " + spec)
-    } catch {
-      case e : StreamValidationException =>
-             if (failOnCheckpointValidation) {
-               throw e
-             } else {
-               warn("Checkpoint stream validation partially failed. Ignoring it because failOnCheckpointValidation=" + failOnCheckpointValidation)
-             }
-      case e1 : Exception => throw e1
+      val checkpointBytes = checkpointEnvelope.getMessage.asInstanceOf[Array[Byte]]
+      val checkpoint = checkpointMsgSerde.fromBytes(checkpointBytes)
+      checkpoints.put(checkpointKey.getTaskName, checkpoint)
     }
+    info(s"Read $numMessagesRead messages from system:$checkpointSystem topic:$checkpointTopic")
+    checkpoints.toMap
   }
 
-  override def register(taskName: TaskName) {
-    debug("Adding taskName " + taskName + " to " + this)
-    taskNames += taskName
-  }
-
-
-  def stop = {
-    synchronized {
-      if (systemProducer != null) {
-        systemProducer.stop
-        systemProducer = null
-      }
+  /**
+    * Returns the oldest available offset for the provided [[SystemStreamPartition]].
+    */
+  private def getOldestOffset(ssp: SystemStreamPartition): String = {
+    val topic: String = ssp.getSystemStream.getStream
+    val partition: Partition = ssp.getPartition
 
-      if (systemConsumer != null) {
-        systemConsumer.stop
-        systemConsumer = null
-      }
+    val metaDataMap = systemAdmin.getSystemStreamMetadata(Collections.singleton(topic))
+    val checkpointMetadata: SystemStreamMetadata = metaDataMap.get(topic)
+    if (checkpointMetadata == null) {
+      throw new SamzaException("Cannot get metadata for system=%s, topic=%s" format(checkpointSystem, topic))
```

Comment:
Nitpick: 
Not your change but:
s/Cannot get/Got null metadata. They have different implications.
Same for partitionMetaData below. Should also not use field name in message (e.g., "partition metadata instead of partitionMetaData")